### ISSUE TITLE
Clarify what is notarized on the ledger

### ISSUE THREAD

**Issue by SteveLasker** [ietf-wg-scitt/draft-ietf-scitt-architecture#321]
From @mcr,

https://mailarchive.ietf.org/arch/msg/scitt/sOGI7xcUaOJx5Zag8uqsygooLrc/

> it seems that the receipt received from service A can easily be added to the statement submitted to service B.  It's unclear to me that the signed statement submitted to service A can then be updated.  Does the Append-Only Log record just the protected and payload of the Sign1, or does it also witness the unprotected header?

We don't seem to clarify what content is used to append to the ledger. Is it just the to-be-signed-bytes, or the entirety of the signed statement, including the unprotected header?

**Comment by achamayou** [comment]
> We don't seem to clarify what content is used to append to the ledger. Is it just the to-be-signed-bytes, or the entirety of the signed statement, including the unprotected header?

I think we should mandate that it's at least (*) the signed statement with an empty unprotected header, because:

1. It's a form of the statement from the issuer that does not contain un-authenticated content.
2. It's convenient for clients/verifiers/tooling to deal with in a way that the TBS is not.

(*) Implementations could decide that they want to allow some unprotected headers in, perhaps as supporting evidence (such as x5chain where it's sent unprotected), and should be encouraged to validate it as much as possible before they do so. Exposing what unprotected headers are counter-signed, if any, could be done in the transparency config, for the benefit of the verifiers.

Allowing unprotected headers wholesale amounts to blindly counter-signing something that hasn't been signed in the first place, and should at the very least be heavily discouraged if not forbidden.

**Comment by SteveLasker** [comment]
Notes from editors meeting:
We should create a PR, for text to be added to the registration section:   [6.2. Registration of Signed Statements](https://ietf-wg-scitt.github.io/draft-ietf-scitt-architecture/draft-ietf-scitt-architecture.html#section-6.2)[Registration of Signed Statements](https://ietf-wg-scitt.github.io/draft-ietf-scitt-architecture/draft-ietf-scitt-architecture.html#name-registration-of-signed-stat)

- Set the value of the unprotected header to an empty map

This would include the protected-header, payload and signature.
We're specifically calling out emptying the unprotected header.

**Comment by robinbryce** [comment]
Per the conversation here https://youtu.be/DK40XJ651Cw I don't object to the requirement that un protected headers be stripped. I can see the could be left to being "client choice", but I do understand the chalenges that could create for scanning the log for the entry. While the client *knows* what was registered, I can see auditors may not.

One concerne I'd like to be clear on in any resulting changes: Did the conversation settle on *permiting* the ledger to include additional content in the actual ledger entry on disc ?

Concrete example from the datatrails ledger we define the merkle leaf of our ledger like this

`H(DOMAIN-SEPARATOR || EXTRABYTES || IDTIMESTAMP || SERIALIZEDBYTES)`

Our *hope* is that SERIALIZEDBYTES is the *result* of the normalization process discussed on the call (regardless of what is decided with respect to the inclusion of unprotected headers), and that its for the TS registration policies to defined additional content. Is this what is intended ?

Additional context:

In our case the log is actually a mog, in that it includes a trie friendly leaf index as well as the more regular merkle tree data. So all material needed to verify the receipt can be obtained (efficiently) from the combination of the original statement, the receipt, and the index entry in the log. This means there is a VDS/ receipts format specific step to assemble the leaf value. 

We do this so that while we may have duplicate SERIALIZEDBYTES, all entries in the ledger leaves have a unique and time sortable component to their pre-image. And one which is established by the process which commits the VDS to persistent storage.

**Comment by rjb4standards** [comment]
Robin,

 

I’m not familiar with the term “mog” can you please provide information to help me understand this?

 

 

Thanks,

 

Dick Brooks

   

Active Member of the CISA Critical Manufacturing Sector, 

Sector Coordinating Council – A Public-Private Partnership

 

 <https://reliableenergyanalytics.com/products> Never trust software, always verify and report! ™

Risk always exists, but trust must be earned and awarded.™ 

https://businesscyberguardian.com/ 

Email: ***@***.***

Tel: +1 978-696-1788

**Comment by achamayou** [comment]
@robinbryce I believe additional content was always allowed, we certainly do it in our implementation, as reflected in the [CCF tree profile](https://github.com/ietf-scitt/draft-birkholz-cose-cometre-ccf-profile/blob/main/draft-birkholz-cose-receipts-ccf-profile.md#transaction-components).

This is (again, as I understand it) orthogonal to _how_ the Signed Statement is normalised prior to inclusion, and whether the unprotected/unsigned bits are included or not.

**Comment by robinbryce** [comment]
> @robinbryce I believe additional content was always allowed, we certainly do it in our implementation, as reflected in the [CCF tree profile](https://github.com/ietf-scitt/draft-birkholz-cose-cometre-ccf-profile/blob/main/draft-birkholz-cose-receipts-ccf-profile.md#transaction-components).
> 
> This is (again, as I understand it) orthogonal to _how_ the Signed Statement is normalised prior to inclusion, and whether the unprotected/unsigned bits are included or not.

Great, that is completely aligned with how I understood things. Thanks!

**Comment by robinbryce** [comment]
Ok, on the specific issue of normalization: my instinct here has always been "don't mess with the bytes" in the strictest sense. A client registering a statement gets to chose and clients may have good reasons for wanting the un protected headers included in the normalised bytes.

The most compeling counter point is that, given a transparent statement with 4 receipts on it that you find under a rock, how would you know which of those receipts contributed pre-image bytes for the others ?

On a transparent statement, receipts are contained in an array

https://ietf-wg-scitt.github.io/draft-ietf-scitt-architecture/draft-ietf-scitt-architecture.html#fig-transparent-statement-edn

Any single registration produces a single receipt.

Can we just say receipts MUST be appended ?

In cases were a client choses not to strip the headers, its then possible to   cdvbfg

**Comment by robinbryce** [comment]
||

**Comment by robinbryce** [comment]
accidental close!

**Comment by robinbryce** [comment]
Last thought: If I wanted to register a statement which represented "this statement was issued and successfully registered, because its got a receipt attached", given the current arch, I would make it the payload and be done I think

**Comment by achamayou** [comment]
> A client registering a statement gets to chose and clients may have good reasons for wanting the un protected headers included in the normalised bytes.

I agree, but the problem with them being unprotected is that there's no telling who set them. It may be the issuer, the client, or someone else. It could be the server implementation.

> The most compeling counter point is that, given a transparent statement with 4 receipts on it that you find under a rock, how would you know which of those receipts contributed pre-image bytes for the others ?

I think the point of stripping the unprotected headers is that you don't need to know, because the receipts don't contribute to the pre-image. All the receipts are for the same pre-image, which allows parallel registration and verification.

**Comment by robinbryce** [comment]
Yeah, I am now leaning quite a lot to remove all. But wondering if there is any support for "remove just receipts" from the headers

**Comment by achamayou** [comment]
I guess the possibilities are:

1. Remove only some things (receipts, ...?)
2. Remove everything
3. Remove everything except some things (x5chain, ...?)

I am not keen on 1. because it amounts to what I perceive as very open-ended acceptance of completely unauthenticated content. A single issuer getting MITM'ed can lead to content unwanted by both the issuer and the ledger making it in. A dispute between the ledger and an issuer about this content also cannot be satisfactorily resolved.

2. seems fine to me, it's what I do! It's also what @OR13 does, if I understand correctly.
3. seems plausible, so long as the content is subject to some kind of constrained validation. Because I can't think of a use case, I'm a bit skeptical though.

**Comment by robinbryce** [comment]
> I agree, but the problem with them being unprotected is that there's no telling who set them. It may be the issuer, the client, or someone else. It could be the server implementation.

And also
>  A single issuer getting MITM'ed can lead to content unwanted by both the issuer and the ledger making it in

Yes totally, but this is why this choice is so impactful it goes to the heart of the trust assumptions:

The *issuer* has signed only the protected headers and the payload. This is not in question here.

The *transparency service* has either:
 A: counter signed (by way of a receipt) only the protected headers and payload
 B: counter signed protected, payload & unprotected(or some redacted subset)
 
For B: whether all or some are signed, the trust model is different for the unprotected headers. The registering entity is the last party in the chain. they are explicitly not necessarily the issuer in the arch.

B: has utility, but the mental model of sig + counter sig is now incomplete
 
 In otherwords, it has quite an impact on the trust model as the unprotected headers are only attested in a meaningful way by the TS. (Granting that the arch does not place trust in the authentication required to call registration endpoints)

**Comment by robinbryce** [comment]
If we do anything other that 2, then the TS is attesting that "this random stuff was included in the statement at time of registration". It would then be a matter for TS registration policies to declare what that means I think ?

**Comment by achamayou** [comment]
> For B: whether all or some are signed, the trust model is different for the unprotected headers. The registering entity is the last party in the chain. they are explicitly not necessarily the issuer in the arch.

Yes, completely agree. I tried to say this in the recording, but the term "counter-signed" proved controversial :)

It seems to me that the unprotected header (minus the receipt) cannot be said to be counter-signed, because it isn't signed to begin with. If the service decides to sign it, then a more obvious and explicit location to do that is in the _protected_ header of the receipt itself. That is very unambiguously "the service says" (as opposed to "the issuer says", which is the signed statement).

I struggle to explain simply what the service signing the unprotected header means, it's something like "the service says that it heard this from a submitter, and the issuer may or may not have meant that".

**Comment by achamayou** [comment]
> If we do anything other that 2, then the TS is attesting that "this random stuff was included in the statement at time of registration". It would then be a matter for TS registration policies to declare what that means I think ?

It's possible to let the registration policy decide, but that makes validating a transparent statement a lot more complicated all of a sudden, because you need to work what the registration policy has decided to leave in or not.

**Comment by JAG-UK** [comment]
I'm not sure which hat I'm wearing for this comment, but here are my thoughts for the consideration of the group.

TL;DR: The unprotected header (UH) MUST be empty when registering a signed statement. 

Why?
As others have argued, the responsibility of the Transparency Service is to faithfully register exactly what the Issuer submits for Registration. Anything else seems strange: the very reason for Issuer signature is to prevent the TS or the channel from tampering or spoofing statements on the Issuer's behalf. 

So then if the Issuer feels the need to put something in the UH then that should be recorded also. Think about why they might do this: to give hints to the Transparency Service? To add tags or something that for whatever reason they want to be able to change and so don't put in the meta map? Well, if these entries are to have any effect *at all*, then those effects might be security relevant to the application layer, or say something critical about the circumstances of Registration, and so not recording them potentially opens an avenue for a dishonest Issuer to smuggle or hide important information away whilst still getting a valid receipt.

BUT the other side of the threat graph also applies: because the UH is not covered by the Sign1 it is possible for a dishonest TS or channel to add things in here, and then record them, and the Issuer would have no recourse to say "that wasn't part of my submission".

And putting these things together one must also conclude that the supposed potential intention of an Issuer to put instructions or metadata or whatever in the UH is unsafe because of this channel/TS threat. And so the assumption must be that no security relevant info can be put in the UH, nor should an Issuer expect that anything in the UH necessarily be processed or actioned by the TS.

At which point it's completely useless and we might as well do away with it.

BUT, there is a stated use for it which remains, which is that on OUTPUT or READ you might have a bunch of useful info stuffed in there, such as SignedStatement + Receipt-in-UH => TransparentStatement. Using the UH as a convenient attached transport for related info that necessarily accrues *after* signing and *after* registration is still pretty neat, and so far the things stuffed in there have their own independent integrity measures so while they're still vulnerable to being removed altogether, there is some trust that can be built there.

**Comment by robinbryce** [comment]
> It seems to me that the unprotected header (minus the receipt) cannot be said to be counter-signed, because it isn't signed to begin with

Absoloutely

>  but that makes validating a transparent statement a lot more complicated

Yes

> (JAG-UK) And so the assumption must be that no security relevant info can be put in the UH

Indeed

A clarification I might then make is "By definition, a transparent statement can not be registered".

If that sort of thing is desired the transparent statement would need to be a payload for a different statement. And as such would be opaque to the protocol

It is implied, but the use of un protected headers to convey receipts has contributed to the need for clarification here I think.

To be clear, I'm supportive of "The unprotected header (UH) MUST be empty when registering a signed statement."

**Comment by robinbryce** [comment]
I also think that

"The unprotected header (UH) MUST be empty when registering a signed statement."

Means that https://ietf-wg-scitt.github.io/draft-ietf-scitt-scrapi/draft-ietf-scitt-scrapi.html#name-register-signed-statement

Should do 400 "invalid request" rather than silently stripping. This means that validation can decode *to check* for the absense of headers, but the exact bytes of the original (valid) request are submited to the ledger. I think has practical benefits for various potential implementation issues.

**Comment by achamayou** [comment]
> "The unprotected header (UH) MUST be empty when registering a signed statement."

That would mean that a client submitting an already transparent statement would need to strip the receipt prior to sending. It's not a major problem (easy enough to ship the CLI tool/utility function), but just to make sure that's what's intended.

Although our implementation strips the UH now, I think I am warming up to the idea of forbidding it altogether, because it's a more definitive solution to canonical serialisation issues, and it ticks the "don't mess with the bytes" box at the same time.

**Comment by JAG-UK** [comment]
> That would mean that a client submitting an already transparent statement would need to strip the receipt prior to sending.

Yes, I can see the practical inconvenience there but I think it's cleaner to just forbid it altogether in the formal spec and make it an SDK problem as you say

**Comment by achamayou** [comment]
@SteveLasker @robinbryce @JAG-UK I realise now, as I am editing to add these changes that:

> Upon registration, the unprotected header MUST be empty, or a 400 is returned via SCRAPI to provide consistency and avoid potential MIM attacks.

substantially conflicts with:

> If `x5t` is included in the protected header, an `x5chain` with a leaf certificate corresponding to the `x5t` value MAY be included in the unprotected header.

To be clear, I am fine with removing this, and mandating that `x5chain` must be protected by being placed directly in the protected header (it is what our implementation does, for the reasons spelled out in #307 and #309). But do we still want to allow implementations that include supporting evidence in unprotected header params, so long as they have a binding to the protected header (`x5t` in this case), _verified_ at registration?

If so, we need more subtle wording, which defines an implementation-specific set of authenticated/bound unprotected header params. Statements containing unprotected header params other than those MUST be rejected then.

What do you think? Is unprotected `x5chain` worth the nuance?

**Comment by robinbryce** [comment]
> If so, we need more subtle wording, which defines an implementation-specific set of authenticated/bound unprotected header params. Statements containing unprotected header params other than those MUST be rejected then.
> 
> What do you think? Is unprotected `x5chain` worth the nuance?

Can we get away with something like just


Upon registration, the unprotected header MUST be empty, of content not otherwise bound explicity by protected headers, or a 400 is returned via SCRAPI to provide consistency and avoid potential MIM attacks.

Or does this force us to enumerate "bound explicitly" in a laborious way ?

**Comment by OR13** [comment]
The unprotected header SHOULD be empty when registering a signed statement.

The server MAY return a 400, and decline to register signed statements with non empty unprotected headers.

**Comment by JAG-UK** [comment]
I like Orie's suggestion. It's very simple and avoids treading on the toes of other bits of the document, or indeed other specs (the unprotected header belongs to COSE - it should already be clear that its contents are not covered by the Issuer Signature).

**Comment by achamayou** [comment]
@OR13 @JAG-UK that's quite a bit more vague than what @robinbryce proposes though, it leaves the door open to services accepting statements with unbound unprotected header params, without specifying if they are to be stripped before inclusion. And I think we've broadly agreed in this thread that this is really undesirable.

I'm otherwise fine with it, in that it allows all desired behaviours (strict rejection of any unprotected header params, limited acceptance of bound unprotected params).

**Comment by OR13** [comment]
@achamayou we can make the SHOULD a MUST, and the MAY a MUST.

If you make these MUSTs, you would not be able to register a receipt in a transparency service.
That use case might not matter, but just for the sake of clarity:

Lets assume you have `receipt.gov.cose` for `software.sbom.cose` and after taking a dependency on the sbom, you want to register both.

You can easily register `software.sbom.cose`, because its just a signed statement.
You cannot register `receipt.gov.cose` without breaking the proof, since the proof is in the unprotected header.

**Comment by achamayou** [comment]
@OR13 I understand the multiple-receipt situation, and I am worried we are [stuck in a loop](https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/321#issuecomment-2552174680). I can see that being able to submit a transparent statement and incrementally pile up receipts is convenient. I can also understand (even I don't follow it) the rationale behind the `x5t`/`x5chain` split.

These are two limited cases where unprotected header parameters are effectively protected, either indirectly, or by another root of trust, such as `receipt.gov.cose` in your example.

The point made in the discussion above is: outside of some exceptions, unprotected header parameters have unknown provenance, since there is no way for the TS to authenticate them. A malicious TS could manufacture some, and an Issuer would not be able to denounce it. A malicious Client sending a statement to the TS on behalf of an Issuer could do the same.

Is it worth removing the ability for auditors to replay registration and end-to-end integrity on statements, for convenience on multiple receipts and detaching `x5chain`?

If we go the SHOULD route, we say that it's up to the TS implementation to decide how to handle unprotected headers, and whether to provide these guarantees. It's not possible to say that a TS implementing the RFC always meets that bar.

I will add, playing devil's advocate here, that even TS implementations that reject non-empty protected headers still have to be careful in how they allow registration policies to be implemented to achieve replayability, because any amount of non-determinism in the policy logic is enough to wreck it.

So I'll go ahead with this change.
